### PR TITLE
Logs: fix exception logging

### PR DESCRIPTION
**Pull request by stsewd**
`record.exc_info` is always `None` for some reason. The only reference to record.exc_info I was able to find on structlog is
https://github.com/hynek/structlog/blob/bcfc7f9e60640c150bffbdaeed6328e582f93d1e/src/structlog/stdlib.py#L308-L325, but I wasn't able to find where that is explicitly used, and in all structlog handlers they take the current exception from the event dict, so..

Closes https://github.com/readthedocs/readthedocs-ops/issues/1262

You can test this locally by adding `NewRelicProcessor()` to any processor, like to `LOGGING.formatters.key_value.processors` (settings/base.py), and log anything with `exc_info=True` or call `log.exception`.

### CODE REVIEW

File: readthedocs/core/logs.py [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
 import warnings
-
 from io import StringIO
 
 import structlog
-from structlog.dev import _pad, plain_traceback
-
 from django_structlog.middlewares.request import RequestMiddleware
+from structlog.dev import _pad, plain_traceback
+from structlog.processors import _figure_out_exc_info
```
Comment (stsewd) [inline readthedocs/core/logs.py:8]: We can just copy this method if we don't want to rely on this private function

https://github.com/hynek/structlog/blob/bcfc7f9e60640c150bffbdaeed6328e582f93d1e/src/structlog/processors.py#L519-L531

File: readthedocs/core/logs.py [hunk 2]
```diff
@@ -69,8 +68,14 @@ def __call__(self, logger, method_name, event_dict):
             "line.number": record.lineno,
         }
 
+        # NOTE: For some reason structlog isn't including
+        # the exception in the record.exc_info attribute,
+        # so we need to get it from the event_dict.
+        exc_info = event_dict.pop("exc_info", None)
```
Comment (benjaoming) [inline readthedocs/core/logs.py:74]: Why is it getting "popped" in stead of getting "getted"? :) If the existing return value (`return event_dict`) was expecting `exc_info`, then it's removed from the dict here..

Comment (stsewd) [reply readthedocs/core/logs.py:74]: So, basically, we are replacing that value with the expected representation of the exception for new relic to render.

Comment (benjaoming) [reply readthedocs/core/logs.py:74]: Nice, thanks for the additional info!

File: readthedocs/core/logs.py [hunk 3]
```diff
@@ -69,8 +68,14 @@ def __call__(self, logger, method_name, event_dict):
...
         if record.exc_info:
             output.update(format_exc_info(record.exc_info))
+        elif exc_info:
+            output.update(format_exc_info(_figure_out_exc_info(exc_info)))
```
Comment (benjaoming) [inline readthedocs/core/logs.py:78]: It's like you can strip out the essential part of the internal structlog function `_figure_out_exc_info` and it makes things less confusing...

```suggestion
        elif isinstance(exc_info, BaseException):
            output.update(
                format_exc_info(
                    (exc_info.__class__, exc_info, getattr(exc_info, "__traceback__"))
                )
            )
        elif isinstance(exc_info, tuple):
            output.update(format_exc_info(exc_info))         
```

Comment (benjaoming) [reply readthedocs/core/logs.py:78]: Maybe this part should also be there: https://github.com/hynek/structlog/blob/bcfc7f9e60640c150bffbdaeed6328e582f93d1e/src/structlog/processors.py#L528-L529

File: readthedocs/core/logs.py (hunk 1 above)
Comment (humitos) [reply readthedocs/core/logs.py:8]: I think it's fine for now.

**Comment by humitos** [comment]
> You can test this locally by adding NewRelicProcessor() to any processor, like to LOGGING.formatters.key_value.processors (settings/base.py), and log anything with exc_info=True or call log.exception.

I would have appreciated a more detailed example to test this 😄 

I've done this:

```diff
diff --git a/readthedocs/settings/base.py b/readthedocs/settings/base.py
index 7fcdc108c..5780a8650 100644
--- a/readthedocs/settings/base.py
+++ b/readthedocs/settings/base.py
@@ -8,7 +8,7 @@ import structlog
 
 from celery.schedules import crontab
 
-from readthedocs.core.logs import shared_processors
+from readthedocs.core.logs import shared_processors, NewRelicProcessor
 from readthedocs.core.settings import Settings
 
 
@@ -922,6 +922,7 @@ class CommunityBaseSettings(Settings):
                     structlog.processors.TimeStamper(fmt='iso'),
                     structlog.stdlib.ProcessorFormatter.remove_processors_meta,
                     structlog.processors.KeyValueRenderer(key_order=['timestamp', 'level', 'event', 'logger']),
+                    NewRelicProcessor(),
                 ],
                 # Allows to add extra data to log entries generated via ``logging`` module
                 # See https://www.structlog.org/en/stable/standard-library.html#rendering-using-structlog-based-formatters-within-logging
```

and then:

```
$ inv docker.manage shell_plus
In [1]: import structlog

In [2]: log = structlog.get_logger('humitos')

In [3]: log.exception('Test')
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.10/logging/__init__.py", line 1100, in emit
    msg = self.format(record)
  File "/usr/lib/python3.10/logging/__init__.py", line 943, in format
    return fmt.format(record)
  File "/usr/local/lib/python3.10/dist-packages/structlog/stdlib.py", line 963, in format
    ed = p(logger, meth_name, ed)
  File "/usr/local/lib/python3.10/dist-packages/structlog/dev.py", line 424, in __call__
    self._exception_formatter(sio, exc_info)
  File "/usr/local/lib/python3.10/dist-packages/structlog/dev.py", line 187, in rich_traceback
    Traceback.from_exception(*exc_info, show_locals=True)
  File "/usr/local/lib/python3.10/dist-packages/rich/traceback.py", line 293, in from_exception
    rich_traceback = cls.extract(
  File "/usr/local/lib/python3.10/dist-packages/rich/traceback.py", line 354, in extract
    exc_type=safe_str(exc_type.__name__),
AttributeError: 'NoneType' object has no attribute '__name__'
Call stack:
  File "/usr/src/app/checkouts/readthedocs.org/manage.py", line 11, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.10/dist-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.10/dist-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.10/dist-packages/django_extensions/management/commands/shell_plus.py", line 125, in run_from_argv
    return super().run_from_argv(argv)
  File "/usr/local/lib/python3.10/dist-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.10/dist-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.10/dist-packages/django_extensions/management/utils.py", line 62, in inner
    ret = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/django_extensions/management/commands/shell_plus.py", line 582, in handle
    runner()
  File "/usr/local/lib/python3.10/dist-packages/django_extensions/management/commands/shell_plus.py", line 404, in run_ipython
    start_ipython(argv=ipython_arguments, user_ns=imported_objects)
  File "/usr/local/lib/python3.10/dist-packages/IPython/__init__.py", line 124, in start_ipython
    return launch_new_instance(argv=argv, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/traitlets/config/application.py", line 1041, in launch_instance
    app.start()
  File "/usr/local/lib/python3.10/dist-packages/IPython/terminal/ipapp.py", line 318, in start
    self.shell.mainloop()
  File "/usr/local/lib/python3.10/dist-packages/IPython/terminal/interactiveshell.py", line 685, in mainloop
    self.interact()
  File "/usr/local/lib/python3.10/dist-packages/IPython/terminal/interactiveshell.py", line 678, in interact
    self.run_cell(code, store_history=True)
  File "/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py", line 2945, in run_cell
    result = self._run_cell(
  File "/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py", line 3000, in _run_cell
    return runner(coro)
  File "/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner
    coro.send(None)
  File "/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py", line 3203, in run_cell_async
    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
  File "/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py", line 3382, in run_ast_nodes
    if await self.run_code(code, result, async_=asy):
  File "/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py", line 3442, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-0f693c4554b8>", line 1, in <module>
    log.exception('Test')
  File "/usr/local/lib/python3.10/dist-packages/structlog/stdlib.py", line 217, in exception
    return self.error(event, *args, **kw)
  File "/usr/local/lib/python3.10/dist-packages/structlog/stdlib.py", line 200, in error
    return self._proxy_to_logger("error", event, *args, **kw)
  File "/usr/local/lib/python3.10/dist-packages/structlog/stdlib.py", line 247, in _proxy_to_logger
    return super()._proxy_to_logger(method_name, event=event, **event_kw)
  File "/usr/local/lib/python3.10/dist-packages/structlog/_base.py", line 206, in _proxy_to_logger
    return getattr(self._logger, method_name)(*args, **kw)
Message: {'exc_info': True, 'event': 'Test', 'logger': 'humitos', 'level': 'error'}
Arguments: ()
```

Am I testing this correctly? I'm not sure what I'm looking for and how to decide whether or not this is working as expected 🤷🏼 

This `AttributeError: 'NoneType' object has no attribute '__name__'` called my attention and it makes me think there is something broken 🙃

File: readthedocs/core/logs.py (hunk 1 above)
Comment (benjaoming) [reply readthedocs/core/logs.py:8]: I found it to be fairly obscure and the private function didn't help to clarify what's going on. I think it'd be nice if @stsewd copies in the original function and adapts it to our needs with some comments on what it's doing :) But I think the Python2-specific parts can be removed.

**Comment by humitos** [comment]
@stsewd I would need some guidance here about how to test this PR locally and also to know what exactly I should be looking for. I raised an exception in the code itself and I don't know how to evaluate if it's working as expected or not; and what's the difference _without_ this PR

**Review by humitos — APPROVED** [review]
Thanks a lot for guiding me about how to test this PR, really appreciated. I just did a small QA again and it's working fine.

**Pull request merged**